designs/cloudstack-multiple-endpoints.md

```diff
@@ -0,0 +1,177 @@
+
+1. Cloudstack endpoint
+2. Cloudstack domain
+3. Cloudstack zone
+4. Cloudstack account
+
+You can find more information about these Cloudstack resources [here](http://docs.cloudstack.apache.org/en/latest/conceptsandterminology/concepts.html#cloudstack-terminology)
+
+### `CloudstackDatacenterConfig` Validation
+
+With the multi-endpoint system for the Cloudstack provider, customers reference a CloudstackMachineConfig and it's created across multiple failure domains. The implication
+is that all the Cloudstack resources such as image, ComputeOffering, ISOAttachment, etc. must be available in *all* the failure domains, or all the Cloudstack endpoints,
+and these resources must be referenced by name, not unique ID. This would mean that for each CloudstackMachineConfig, we have to make sure that all the prerequisite
+Cloudstack resources are available in all the Cloudstack API endpoints. 
+
+In practice, the pseudocode would look like:
+
+for failureDomain in failureDomains:
+  for machineConfig in machineConfigs:
+    validate resource presence with the failureDomain's instance of the CloudMonkey executable
+
+### Cloudstack credentials
+
+
+In a multi-endpoint Cloudstack cluster, each endpoint may have its own credentials. We propose that Cloudstack credentials will be passed in via environment variable in the same way as they are currently,
+only as a list corresponding to failure domains. Currently, these credentials are passed in via environment variable, which contains a base64 encoded .ini file that looks like
+
+```
+[Global]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.1:8080/client/api
+```
+
+We would propose an extension of the above input mechanism so the user could provide credentials across multiple Cloudstack API endpoints like
+
+```
+[FailureDomain1]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.1:8080/client/api
+
+[FailureDomain2]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.2:8080/client/api
+
+[FailureDomain3]
+api-key    = redacted
+secret-key = redacted
+api-url    = http://172.16.0.3:8080/client/api
+
+...
+```
+
+We are also exploring converting the ini file to a yaml input file which contains a list of credentials and their associated endpoints. Either way, this environment variable would
+be passed along to CAPC and used by the CAPC controller just like it is currently.
+
+### Backwards Compatibility
+
+Our customer currently has clusters running with the old resource definition. In order to support backwards compatibility in the CloudstackDatacenterConfig resource, we can
+1. Make all the fields optional and see if customers have the old fields set or the new ones
+2. Introduce an eks-a version bump with conversion webhooks
+
+Between these two approaches, I would take the first and then deprecate the legacy fields in a subsequent release to simplify the code paths.
```

Comment:
I think we can still throw a warning if new clusters are created with the old fields.